Experiments: Updates custom logging page + addresses Jira ticket (#1269)
wandb/docs: content/guides/models/track/log/customize-logging-axes.md

title: Customize log axes
---

> Set a custom x-axis when you log metrics to W&B. By default, W&B logs metrics as a function of *steps*. A step corresponds to the number of times you call `wandb.log()`. 

Reviewer: Not sure about "as a function of steps". Is it better to. just say that the X-axis defaults to step?

> For example, suppose you log a metric in a for loop that iterates 10 times. W&B logs the metric against the step number, which is incremented by 1 each time you log a metric.

Reviewer: ```suggestion
For example, suppose you log a metric in a Python `for` loop that iterates 10 times. W&B logs the metric against the step number, which is incremented by 1 each time you log a metric.
```

The code might look similar to the following:

```python
import wandb

with wandb.init() as run:
  for i in range(10):
    log_dict = {
        "validation/loss": 1/(i+1)   
    }
    run.log(log_dict)
```

> The previous code snippet generates a line plot panel in the W&B UI that shows the `validation/loss` metric as a function of the step number. 

Reviewer: Similar feedback to the above

{{< img src="/images/experiments/standard_axes.png" alt="Line plot panel that uses default x-axis. Each step in monotonically increases by 1." >}}

> However, there are cases where you might want to log metrics against a different x-axis. In this case, use the [`define_metric`]({{< relref "ref/python/run/#define_metric" >}}) method to define a custom x-axis. 

Reviewer: ```suggestion
In certain situations, it makes more sense to log metrics against a different X-axis. For example, <some example>. Use the [`define_metric`]({{< relref "ref/python/run/#define_metric" >}}) method to define a custom x-axis. 
```
If you take this suggestion, replace `<some example>` with something. :)

> Specify the name of the metric you want to create a custom x-axis for with the `name` parameter. Then, set the `step_metric` parameter to the name of the metric you want to use as the x-axis.

Reviewer: It may be better to move this up into the previous para.

> The following code snippet demonstrates how to set a custom x-axis metric called `"custom_step"` and log the `"validation_loss"` metric against it. With each iteration in the for loop, the `"custom_step"` is set to `i**2`:

Reviewer: Do you need to have the quotes here? It might be confusing to leave them in.

```python
import wandb

with wandb.init() as run:
    # define which metrics will be plotted against it
    run.define_metric(name = "validation_loss", step_metric = "custom_step")

    for i in range(10):
        log_dict = {
            "custom_step": i**2,
            "validation_loss": 1 / (i + 1),
        }
        run.log(log_dict)
```

> Within the W&B UI, you can see that the `validation_loss` is plotted against `custom_step`. Note how the x-axis is set to the square of the for loop index `i`.

Reviewer: ```suggestion
Within the W&B UI,  `validation_loss` is plotted against `custom_step`. Note how the x-axis is set to the square of the for loop index `i`.
```
"the square of the for loop index" seems like a confusing way to explain this.

title: Customize log axes
---

Set a custom x-axis when you log metrics to W&B. By default, W&B logs metrics as *steps*. Each step corresponds to a `wandb.log()` API call. 

> For example, suppose you have a `for` loop that loops 10 times (see the following code block). In each `for` loop, you log a metric called `validation_loss` with `wandb.log()`. W&B increments the step number by 1 each time you log that metric. 

Reviewer: ```suggestion
For example, the following script has a `for` loop that iterates 10 times. In each iteration, the script logs a metric called `validation_loss` and increments the step number by 1.
```

```python
import wandb

with wandb.init() as run:
  # range function creates a sequence of numbers from 0 to 9
  for i in range(10):
    log_dict = {
        "validation_loss": 1/(i+1)   
    }
    run.log(log_dict)
```

> If you navigate to your project's workspace, you can see that the `validation_loss` metric is plotted against the step x-axis. Each step in the for loop monotonically increases by 1, so the x-axis shows the step numbers 0, 1, 2, ..., 9. (The Python `range()` function creates a sequence of numbers from 0 to 9.)

Reviewer: ```suggestion
In the project's workspace, the `validation_loss` metric is plotted against the `step` x-axis, which increments by 1 each time `run.log()` is called.
```
You already have a comment in the script that explains `range()` if it's not assumed knowledge for Python developers.

> {{< img src="/images/experiments/standard_axes.png" alt="Line plot panel that uses default x-axis. Each step in monotonically increases by 1." >}}

Reviewer: ```suggestion
{{< img src="/images/experiments/standard_axes.png" alt="Line plot panel that uses `step` as the x-axis." >}}
```
The alt text doesn't need to function as the actual documentation. :)

> In certain situations, it makes more sense to log metrics against a different x-axis. For example, you may want a logarithmic x-axis. Use the [`define_metric()`]({{< relref "ref/python/run/#define_metric" >}}) method to define a custom x-axis based on a metric you log.

Reviewer: ```suggestion
In certain situations, it makes more sense to log metrics against a different x-axis such as a logarithmic x-axis. Use the [`define_metric()`]({{< relref "ref/python/run/#define_metric" >}}) method to use any metric you log  as a custom x-axis.
```
I got confused between line 29 and 31. Line 29 talks about custom X-axis, then line 31 begins by talking about a custom Y-axis before returning to the custom X-axis. Did something get mixed up along the way?

Author: Funny. So that y-value formula was there to make the plots look nice. I can see how this might be confusing. I updated it , let me know if it makes more sense.

Specify the metric that you want to appear as the y-axis with the `name` parameter. The `step_metric` parameter specifies the metric you want to use as the x-axis. When you log a metric, specify a value for both the x-axis and the y-axis as key-value pairs in a dictionary. 

Copy and paste the following code snippet to set a custom x-axis metric. Replace the values within `<>` with your own values:

```python
import wandb

custom_step = "<custom_step>"  # custom x-axis
metric_name = "<metric>"  # metric to log against custom x-axis

with wandb.init() as run:
    # define step metric (x-axis) and the metric to log against it (y-axis)
    run.define_metric(name = metric_name, step_metric = custom_step)

    for i in range(10):
        log_dict = {
            custom_step : int,  # Value of x-axis
            metric_name : int,  # Value of y-axis
        }
        run.log(log_dict)
```

> As an example, the following code snippet demonstrates how to create a custom x-axis using a metric called `x_axis_squared` and a y-axis called `validation_loss`. The value of the custom x-axis is the square of the for loop index `i` (`i**2`). The `validation_loss` metric is logged as `1 / (i + 1)`.

Reviewer: ```suggestion
As an example, the following code snippet creates a custom x-axis using the metric `x_axis_squared` and the y-axis using the metric `validation_loss`. The value of the custom x-axis is the square of the for loop index `i` (`i**2`). The `validation_loss` metric is logged as `1 / (i + 1)`.
```
